library/src/main/scala/JavaCodeGen.scala

```diff
@@ -0,0 +1,214 @@
+
+  private def genFile(d: Definition) = {
+    val fileName = d.name + ".java"
+    d.namespace map (ns => new File(ns.replace(".", File.separator), fileName)) getOrElse new File(fileName)
+  }
+
+  private def genFields(fields: List[Field]) = fields map genField mkString EOL
+  private def genField(f: Field) =
+    s"""${genDoc(f.doc)}
+       |private ${genRealTpe(f.tpe)} ${f.name};""".stripMargin
+
+  def isPrimitive(tpe: TpeRef) = !tpe.repeated && ! tpe.lzy && tpe.name != boxedType(tpe.name)
+
+  private def boxedType(tpe: String): String = tpe match {
+    case "boolean" => "Boolean"
+    case "byte"    => "Byte"
+    case "char"    => "Character"
+    case "float"   => "Float"
+    case "int"     => "Integer"
+    case "long"    => "Long"
+    case "short"   => "Short"
+    case "double"  => "Double"
+    case other     => other
+  }
+
+  private def genRealTpe(tpe: TpeRef): String = tpe match {
+    case TpeRef(name, true, true)   => s"Lazy<$name[]>"
+    case TpeRef(name, true, false)  => s"Lazy<${boxedType(name)}>"
+    case TpeRef(name, false, true)  => s"$name[]"
+    case TpeRef(name, false, false) => name
+  }
+
+  private def genAccessors(fields: List[Field]) = fields map genAccessor mkString EOL
+  private def genAccessor(field: Field) = {
+    val accessCode =
+      if (field.tpe.lzy) s"this.${field.name}.get()"
+      else s"this.${field.name}"
+
+    // We don't use `genRealTpe` here because the field will be evaluated
+    // if it is lazy.
+    val tpeSig =
+      if (field.tpe.repeated) s"${field.tpe.name}[]"
+      else field.tpe.name
+
+    s"""public $tpeSig ${field.name}() {
+       |    return $accessCode;
+       |}""".stripMargin
+  }
+
+  private def genConstructors(cl: ClassLike, parent: Option[Protocol], superFields: List[Field]) =
+    perVersionNumber(superFields ++ cl.fields) { (provided, byDefault) =>
+      val ctorParameters = provided map (f => s"${genRealTpe(f.tpe)} _${f.name}") mkString ", "
+      val superFieldsValues = superFields map {
+        case f if provided contains f  => s"_${f.name}"
+        case f if byDefault contains f => f.default getOrElse sys.error(s"Need a default value for field ${f.name}.")
+      }
+      val superCall = superFieldsValues.mkString("super(", ", ", ");")
+      val assignments = cl.fields map {
+        case f if provided contains f  => s"${f.name} = _${f.name};"
+        case f if byDefault contains f => f.default map (d => s"${f.name} = $d;") getOrElse sys.error(s"Need a default value for field ${f.name}.")
+      } mkString EOL
+
+      s"""public ${cl.name}($ctorParameters) {
+         |    $superCall
+         |    $assignments
+         |}""".stripMargin
+    } mkString (EOL + EOL)
+
+  private def genEquals(cl: ClassLike, superFields: List[Field]) = {
+    val allFields = superFields ++ cl.fields
+    val body =
+      if (allFields exists (_.tpe.lzy)) {
+        "return this == obj; // We have lazy members, so use object identity to avoid circularity."
```

Comment:
I've added the same check for `toString`, plus a comment in the generated code for `hashCode` and `toString`.